File: benchmarks/executable/integer_saturating_conversion.c
// C23 reference for representative Restaurant integer saturating conversions.
// Expected exit: 0
// Expected stdout:
// ss -128/7/127; us 7/127/127; su 0/200/255; uu 7/255/255; UInt->Int 9223372036854775807

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile int16_t signed16[] = {-129, 7, 128, -1, 200, 256};
static volatile uint16_t unsigned16[] = {7u, 127u, 128u, 255u, 256u};
static volatile uint64_t unsigned64_maximum = UINT64_MAX;

static int8_t saturate_signed_to_signed_i8(int16_t value) {
  if (value < INT8_MIN) return INT8_MIN;
  if (value > INT8_MAX) return INT8_MAX;
  return (int8_t)value;
}

static int8_t saturate_unsigned_to_signed_i8(uint16_t value) {
  return value > (uint16_t)INT8_MAX ? INT8_MAX : (int8_t)value;
}

static uint8_t saturate_signed_to_unsigned_u8(int16_t value) {
  if (value < 0) return 0u;
  if (value > UINT8_MAX) return UINT8_MAX;
  return (uint8_t)value;
}

static uint8_t saturate_unsigned_to_unsigned_u8(uint16_t value) {
  return value > UINT8_MAX ? UINT8_MAX : (uint8_t)value;
}

static int64_t saturate_uint_to_int(uint64_t value) {
  return value > (uint64_t)INT64_MAX ? INT64_MAX : (int64_t)value;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  if (printf("ss %d/%d/%d; us %d/%d/%d; su %u/%u/%u; uu %u/%u/%u; "
             "UInt->Int %" PRId64 "\n",
             (int)saturate_signed_to_signed_i8(signed16[0]),
             (int)saturate_signed_to_signed_i8(signed16[1]),
             (int)saturate_signed_to_signed_i8(signed16[2]),
             (int)saturate_unsigned_to_signed_i8(unsigned16[0]),
             (int)saturate_unsigned_to_signed_i8(unsigned16[1]),
             (int)saturate_unsigned_to_signed_i8(unsigned16[2]),
             (unsigned)saturate_signed_to_unsigned_u8(signed16[3]),
             (unsigned)saturate_signed_to_unsigned_u8(signed16[4]),
             (unsigned)saturate_signed_to_unsigned_u8(signed16[5]),
             (unsigned)saturate_unsigned_to_unsigned_u8(unsigned16[0]),
             (unsigned)saturate_unsigned_to_unsigned_u8(unsigned16[3]),
             (unsigned)saturate_unsigned_to_unsigned_u8(unsigned16[4]),
             saturate_uint_to_int(unsigned64_maximum)) < 0)
    return 1;
  return 0;
}

File: benchmarks/executable/integer_saturating_conversion.rs
// Rust 2024 reference for representative Restaurant integer saturating conversions.
// Expected exit: 0
// Expected stdout:
// ss -128/7/127; us 7/127/127; su 0/200/255; uu 7/255/255; UInt->Int 9223372036854775807

use std::hint::black_box;

fn saturate_signed_to_signed_i8(value: i16) -> i8 {
    value.clamp(i8::MIN as i16, i8::MAX as i16) as i8
}

fn saturate_unsigned_to_signed_i8(value: u16) -> i8 {
    value.min(i8::MAX as u16) as i8
}

fn saturate_signed_to_unsigned_u8(value: i16) -> u8 {
    value.clamp(0, u8::MAX as i16) as u8
}

fn saturate_unsigned_to_unsigned_u8(value: u16) -> u8 {
    value.min(u8::MAX as u16) as u8
}

fn saturate_uint_to_int(value: u64) -> i64 {
    value.min(i64::MAX as u64) as i64
}

fn main() {
    println!(
        "ss {}/{}/{}; us {}/{}/{}; su {}/{}/{}; uu {}/{}/{}; UInt->Int {}",
        saturate_signed_to_signed_i8(black_box(-129_i16)),
        saturate_signed_to_signed_i8(black_box(7_i16)),
        saturate_signed_to_signed_i8(black_box(128_i16)),
        saturate_unsigned_to_signed_i8(black_box(7_u16)),
        saturate_unsigned_to_signed_i8(black_box(127_u16)),
        saturate_unsigned_to_signed_i8(black_box(128_u16)),
        saturate_signed_to_unsigned_u8(black_box(-1_i16)),
        saturate_signed_to_unsigned_u8(black_box(200_i16)),
        saturate_signed_to_unsigned_u8(black_box(256_i16)),
        saturate_unsigned_to_unsigned_u8(black_box(7_u16)),
        saturate_unsigned_to_unsigned_u8(black_box(255_u16)),
        saturate_unsigned_to_unsigned_u8(black_box(256_u16)),
        saturate_uint_to_int(black_box(u64::MAX)),
    );
}
